### DISCUSSION TITLE
How to set up symlink on Windows (and several other questions)?

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [ccache/ccache#797] [Support]
I will just get straight to the point. So what Ive tried so far is:

mklink /J C:\Program Files\LLVM\bin C:\Program Files\ccache

which tells me that I cant create symlinks for existing directories.

This is kinda what I expected to see tbh. Im not sure if I understand the user manual correctly, if Im supposed to have a link pointing towards the ccashe directory where I would normally have my compiler, then where do I put my compiler? And how does ccashe know how to find the compiler as well as what compiler to use?

Im also curious what works best on Windows, ccashe with Clang or cl.exe?

I should also mention that the reason for the interested of using the symlink solution is that Im using a non-source build of Unreal Engine (and as far as I know there are no ways of replacing the by default supported compilers without modyfying the source code). If you have any better solutions to involving ccashe in Unreals build process then please do share

Im also curious. Does ccashe support reusing just portions of each compiled source file (cpp/h)? Or will it only reuse entire files (and when a file doesnt match, it wont attempt reusing at all)?

Thanks in advance :)

**Comment by jrosdahl** [comment]
I assume you're referring to the "symlinks method" (masquerading) in the [_Run modes_ section of the manual](https://ccache.dev/manual/4.2.html#_run_modes). That section is written for Unix systems, not Windows. I have no idea if the symlinks method works on Windows.

> Im not sure if I understand the user manual correctly, if Im supposed to have a link pointing towards the ccashe directory where I would normally have my compiler, then where do I put my compiler?

The symlink should not be for directories but for executables. The idea is that the ccache executable is found first in PATH but named as the compiler. Ccache will then (when invoked with the compiler name) notice this and execute the real compiler under the hood by locating it from PATH.

For Windows, I guess it should be something like this (if it works at all; I have never run ccache on Windows myself):

```
mklink "C:\some\place\clang.exe" "C:\Program Files\ccache\ccache.exe"
```

If `%PATH%` looks like `C:\some\place;C:\Program Files\LLVM\bin;...` then `C:\some\place\clang.exe` will be executed if you run `clang` in the command prompt, and since that actually will be `C:\Program Files\ccache\ccache.exe` ccache will then find and execute `C:\Program Files\LLVM\bin\clang.exe`.

> And how does ccashe know how to find the compiler as well as what compiler to use?

The compiler is found from PATH. The compiler to use is what ccache is called as (`clang.exe` in the example above).

> Im also curious what works best on Windows, ccashe with Clang or cl.exe?

Ccache doesn't support MSVC (`cl.exe`). What made you think it did? The reason I ask is to understand if we could improve documentation in some way. (Ccache also doesn't support `clang-cl` (Clang in MSVC compatibility mode).)

> Im also curious. Does ccashe support reusing just portions of each compiled source file (cpp/h)? Or will it only reuse entire files (and when a file doesnt match, it wont attempt reusing at all)?

Only entire files.

**Reply by ItsCubeTime** [reply]
Hey and thanks for the reply!

> I assume you're referring to the "symlinks method"

Yep

> The symlink should not be for directories but for executables.

Yeah that cleared things up heh. Its my first time using Symlinks. So basically any symlink will always appear before the file/directory its pointing towards in PATH?

> Ccache doesn't support MSVC (cl.exe). What made you think it did? The reason I ask is to understand if we could improve documentation in some way. (Ccache also doesn't support clang-cl (Clang in MSVC compatibility mode).)

My mistake, I probably read ["May get partial B-level support in the future. See pull request #506."](https://ccache.dev/platform-compiler-language-support.html)] a little fast.

**Reply by ItsCubeTime** [reply]
Is there a way that I can know if ccashe is being used or not once Ive setup my symlink?

**Reply by ItsCubeTime** [reply]
![image](https://user-images.githubusercontent.com/20190653/107398785-46630a80-6b00-11eb-8e9b-0b318824b039.png)

In my case after creating the symlink and adding the ccashe directory to the top of PATH Im not getting any return. Am I supposed not to add the symlink to path? (Or could this be a possible incompatability issue?)

EDIT: Indeed I suspect there might be some Windows compatability issue here, at one point (I took a 20 minute dinner break earlier after running those cmds on the above screenshot), when I was back at my pc I saw quite high CPU usage for no apparent reason, making me think maybe ccashe might have been calling itself simultaneously instead of the actual Clang compiler. Setting the "path" variable in ccache.conf does the trick and I get greeted by clangs help page through the symlink.



EDIT 2: Right, so it turns out that Unreal Engine uses clang-cl.exe and not clang.exe. So I suppose theres nothing I can do from here then (other than go back to using Clang without ccashe)? I already tried setting compiler=clang in the config file, however when I do that I get thrown with a bunch of errors by Unreals Build Tool.

This is the UBT "Live++" log when forcing regular clang compiler if you are interested btw: https://pastebin.com/b6UBs8YR

**Reply by jrosdahl** [reply]
> So basically any symlink will always appear before the file/directory its pointing towards in PATH?

Yes, the directory with a symlink to ccache (but with the compiler's name) needs to appear before the directory with the actual compiler in PATH.

> Is there a way that I can know if ccashe is being used or not once Ive setup my symlink?

Check the output of `ccache -s`.

**Reply by ItsCubeTime** [reply]
Thanks.

I did some more testing and now when calling clang or clang-cl from terminal using a symlink pointing towards ccache, everything runs fine however when attempting to run it from Unreal I see this:

```ERROR: Unable to create child process```



```
Building BullCowGameEditor...
Using Clang 0.0.0 compiler (C:\Program Files\LLVM) with Visual Studio 2019 14.28.29335 runtime (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333) and Windows 10.0.19041.0 SDK (C:\Program Files (x86)\Windows Kits\10).
[Upgrade]
[Upgrade] Using backward-compatible build settings. The latest version of UE4 sets the following values by default, which may 
require code changes:
[Upgrade]     bLegacyPublicIncludePaths = false                 => Omits subfolders from public include paths to reduce compiler command line length. (Previously: true).
[Upgrade]     ShadowVariableWarningLevel = WarningLevel.Error   => Treats shadowed variable warnings as errors. (Previously: WarningLevel.Warning).
[Upgrade]     PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs   => Set in build.cs files to enables IWYU-style PCH model. See 
https://docs.unrealengine.com/en-US/Programming/BuildTools/UnrealBuildTool/IWYU/index.html. (Previously: PCHUsageMode.UseSharedPCHs).
[Upgrade] Suppress this message by setting 'DefaultBuildSettings = BuildSettingsVersion.V2;' in BullCowGameEditor.Target.cs, and explicitly overriding settings that differ from the new defaults.
[Upgrade]
Building 15 actions with 4 processes...
  [1/15] SharedPCH.Engine.cpp
  ERROR: Unable to create child process
  [2/15] Default.rc2
The terminal process "C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -Command Engine\Build\BatchFiles\Build.bat BullCowGameEditor Win64 Development 'C:\Users\Oliver\Documents\Unreal Projects\BullCowGame-starter-kit\BullCowGame.uproject' -waitmutex" terminated with exit code: 1.

Terminal will be reused by tasks, press any key to close it.

```

I suppose you dont know what might be the cause here?

I should also mention that Unreal Build Tools typically detect the apprioporite Clang version. In order to get Unreal to even attempt using the symlinks I had to move them to ```...\program files\LLVM\bin```

**Reply by jrosdahl** [reply]
> I suppose you dont know what might be the cause here?

Nope.

**Reply by ItsCubeTime** [reply]
> > I suppose you dont know what might be the cause here?
> 
> Nope.

Right, it doesn't seem like Im getting this to quite work for my purpose here then (Unreal Engine uses overly cluttered headers and seem to recompile files when its not quite necessary from time to time, giving quite unnecessarily long compile times).

I did get it to compile through ccache.exe right about now. However ccache seems not to do any actual caching (I suppose this is default behaviour when clang-cl.exe is called through ccache?).

I will make it a feature request to support ```clang-cl.exe```. Really interesting project!

Och kul att träffa på någon från samma stad, råkar också befinna mig i Linköping. :)
